source/server/admin/grouped_stats_request.cc

```diff
@@ -0,0 +1,156 @@
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_index_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> GroupedStatsRequest::saveMatchingStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if unused
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
+        return true;
+      }
+    } else if (params_.re2_filter_ != nullptr &&
+               !re2::RE2::PartialMatch(stat->name(), *params_.re2_filter_)) {
+      return true;
+    }
+
+    // capture stat by either adding to a pre-existing variant or by creating a new variant
+    std::string tag_extracted_name = global_symbol_table_.toString(stat->tagExtractedStatName());
+    auto [iterator, inserted] = stat_map_.try_emplace(
+        tag_extracted_name, std::vector<Stats::RefcountPtr<StatType>>({stat}));
+    if (!inserted) {
+      absl::get<std::vector<Stats::RefcountPtr<StatType>>>(iterator->second).emplace_back(stat);
+    }
+    return true;
+  };
+}
+
+Stats::IterateFn<Stats::TextReadout> GroupedStatsRequest::saveMatchingStatForTextReadout() {
+  return saveMatchingStat<Stats::TextReadout>();
+}
+
+Stats::IterateFn<Stats::Gauge> GroupedStatsRequest::saveMatchingStatForGauge() {
+  return saveMatchingStat<Stats::Gauge>();
+}
+
+Stats::IterateFn<Stats::Counter> GroupedStatsRequest::saveMatchingStatForCounter() {
+  return saveMatchingStat<Stats::Counter>();
+}
+
+Stats::IterateFn<Stats::Histogram> GroupedStatsRequest::saveMatchingStatForHistogram() {
+  return saveMatchingStat<Stats::Histogram>();
+}
+
+template <class SharedStatType>
+void GroupedStatsRequest::renderStat(const std::string& name, Buffer::Instance& response,
+                                     const StatOrScopes& variant) {
+  auto prefixed_tag_extracted_name = prefixedTagExtractedName<SharedStatType>(name);
+  if (prefixed_tag_extracted_name.has_value()) {
+    PrometheusStatsRender* const prometheus_render =
+        dynamic_cast<PrometheusStatsRender*>(render_.get());
+    // increment stats count
+    ++phase_stat_count_;
+
+    // sort group
+    std::vector<SharedStatType> group = absl::get<std::vector<SharedStatType>>(variant);
+    std::sort(group.begin(), group.end(),
+              [this](const Stats::RefcountPtr<Stats::Metric>& stat1,
+                     const Stats::RefcountPtr<Stats::Metric>& stat2) -> bool {
+                return global_symbol_table_.lessThan(stat1->statName(), stat2->statName());
+              });
+
+    // render group
+    StatOrScopesIndex index = static_cast<StatOrScopesIndex>(variant.index());
+    std::string type = (index == StatOrScopesIndex::Counter) ? "counter" : "gauge";
```

Comment:
ASSERT that it's a gauge if not a counter? It looks like you can get here with a text-readout; is that right?

This may be correct but it could use some comments.

Reply:
Yep, you can reach this code with a text-readout, which are considered of type `gauge` when rendered (see https://github.com/envoyproxy/envoy/blob/9d9940b49f642fcf483ac8ff387ecb521dcf94fc/source/server/admin/prometheus_stats.cc#L278-L280).

I will add a comment to explain that.